### PR TITLE
Deppy Framework (6/7): dimacs deppy command

### PR DESCRIPTION
**Pull request by perdasilva**
This PR adds a deppy command for solving arbitrary sat problems described in dimacs format

**Note**: to make dimacs work I had to change the definition of sat.Constraint to export the previously unexported methods (e.g. anchor, apply, etc.), so that I could define my own constraint type. I think we need to figure out a way to build these constraints with And, Or, Not top level constraints.

### CODE REVIEW

File: cmd/dimacs/cmd.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+	"github.com/operator-framework/deppy/internal/constraints"
+	"github.com/operator-framework/deppy/internal/entitysource"
+	"github.com/operator-framework/deppy/internal/solver"
+)
+
+var Cmd = &cobra.Command{
+	Use:   "solve <path>",
+	Short: "Solves a sat problem given in dimacs format",
+	Long:  `Solves a sat problem given in dimacs format"`,
```
Comment (oceanc80) [inline cmd/dimacs/cmd.go:19]: Extra `"`

File: cmd/dimacs/cmd.go [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+		return nil
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return solve(args[0])
+	},
+}
+
+func solve(path string) error {
+	// open dimacs file
+	dimacsFile, err := os.Open(path)
```
Comment (oceanc80) [inline cmd/dimacs/cmd.go:36]: Need to close `dimacsFile`

Comment (perdasilva) [reply cmd/dimacs/cmd.go:39]: Nice, one! Fixed

File: cmd/dimacs/cmd.go (hunk 1 above)
Comment (perdasilva) [reply cmd/dimacs/cmd.go]: Fixed

File: cmd/dimacs/cmd.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+func NewDimacsCommand() *cobra.Command {
+	return &cobra.Command{
+		Use:   "solve <path>",
+		Short: "Solves a sat problem given in dimacs format",
+		Long:  `Solves a sat problem given in dimacs format`,
+		Args: func(cmd *cobra.Command, args []string) error {
+			if err := cobra.ExactArgs(1)(cmd, args); err != nil {
+				return err
+			}
```
Comment (timflannagan) [inline cmd/dimacs/cmd.go:25]: nit: I'd move the file stating operation to a PreRun, or in the RunE flow, which means the args field looks like the following:

```go
Args: cobra.ExactArgs(1),
RunE: ...
```

File: cmd/dimacs/cmd.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	defer dimacsFile.Close()
+
+	dimacs, err := NewDimacs(dimacsFile)
+	if err != nil {
+		return fmt.Errorf("error parsing dimacs file (%s): %w", path, err)
+	}
+
+	// build solver
+	so, err := solver.NewDeppySolver(entitysource.NewGroup(NewDimacsEntitySource(dimacs)), constraints.NewConstraintAggregator(NewDimacsConstraintGenerator(dimacs)))
+
```
Comment (timflannagan) [inline cmd/dimacs/cmd.go:52]: nit: remove the newline here?

File: cmd/dimacs/cmd.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		return err
+	}
+
+	// get solution
+	solution, err := so.Solve(context.Background())
+	if err != nil {
+		fmt.Printf("no solution found: %s\n", err)
```
Comment (timflannagan) [inline cmd/dimacs/cmd.go:60]: Do we want/need a custom error here to delineate between legitimate errors encountered while performing SAT solving, vs. the inability to produce a solution?

File: cmd/dimacs/cmd.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import (
+	"context"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+
+	"github.com/operator-framework/deppy/internal/entitysource"
+
```
Comment (timflannagan) [inline cmd/dimacs/cmd.go:12]: nit: remove newline here?

File: cmd/dimacs/cmd.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	"github.com/operator-framework/deppy/internal/constraints"
+	"github.com/operator-framework/deppy/internal/solver"
+)
+
+func NewDimacsCommand() *cobra.Command {
+	return &cobra.Command{
+		Use:   "solve <path>",
+		Short: "Solves a sat problem given in dimacs format",
+		Long:  `Solves a sat problem given in dimacs format`,
```
Comment (timflannagan) [inline cmd/dimacs/cmd.go:21]: TODO: Add an example here?

**Review by timflannagan — APPROVED** [review]
I'm not entirely convinced on the dimacs format, but if people fine it valuable for iterating, or being able to showcase some quick deppy solving mechanisms, then this is cool with me.

Had some nits, but nothing blocking.

**Review by akihikokuroda — APPROVED** [review]
LGTM.  I haven't looked at the dimacs.  I'll play with it later.  It doesn't hurt merging this.

File: cmd/dimacs/cmd.go (hunk 5 above)
Comment (perdasilva) [reply cmd/dimacs/cmd.go:60]: I'll leave this as a question

File: cmd/dimacs/cmd.go (hunk 3 above)
Comment (perdasilva) [reply cmd/dimacs/cmd.go]: Fixed

File: cmd/dimacs/cmd.go (hunk 6 above)
Comment (perdasilva) [reply cmd/dimacs/cmd.go]: fixed

**Comment by akihikokuroda** [comment]
I've just tried with a sample file.  I got 
```
Error: error parsing dimacs file (./dimacs2.txt): invalid format: number of clauses in header differ from the total number of clauses
```
Am I doing something wrong?

**Comment by perdasilva** [comment]
@akihikokuroda can you paste it here? could be there's a bug in the parser =P

**Comment by akihikokuroda** [comment]
Sorry, it's my bad.  I found that my file doesn't have the newline at the end of the last line.

**Pull request merged**